sycl/doc/extensions/usm/usm.adoc

```diff
@@ -0,0 +1,474 @@
+
+== SYCL Memory Management
+This section describes new properties and routines for pointer-based memory management interfaces in SYCL.  These routines augment, rather than replace, the existing buffer-based interfaces in SYCL 1.2.1.  
+
+=== Unified Addressing
+Unified Addressing guarantees that all devices will use a unified address space.  Pointer values in the unified address space will always refer to the same location in memory.  The unified address space encompasses the host and one or more devices.  Note that this does not require addresses in the unified address space to be accessible on all devices, just that pointer values will be consistent.  
+[cols="^25,^15,60",options="header"]
+
+=== Unified Shared Memory
+Unified Shared Memory (USM) is a capability that, when available, provides the ability to create allocations that are visible to both host and device(s).  USM builds upon Unified Addressing to define a shared address space where pointer values in this space always refer to the same location in memory.  USM defines multiple tiers of increasing capability described in the following sections. 
+
+NOTE: All utility functions described below are located in the `sycl` namespace unless otherwise indicated.
+
+=== Explicit USM
+Explict USM defines capabilities for explicitly managing device memory.  Programmers directly allocate device memory, and data must be explicitly copied between the host and a device.  Device allocations are obtained through a SYCL device allocator instead of the system allocator.  Device allocations are not accessible on the host, but the pointer values remain consistent on account of Unified Addressing.  Greater detail about how allocations are used is described by the following tables.
+
+[cols="^25,75",options="header"]
+|===
+|Allocation Type |Description
+|`device`
+|Allocations in device memory that are *not* accessible by the host.
+|===
+
+[cols="6*^",options="header",stripes=none]
+|===
+|Allocation Type |Initial Location |Accessible By | |Migratable To |
+.3+^.^|`device`
+.3+^.^|`device`
+|`host`
+|{n}
+|`host`
+|{n}
+
+|`device`
+|{y}
+|`device`
+|N/A
+
+|Another `device`
+|Optional (P2P)
+|Another `device`
+|{n}
+
+|===
+
+==== Explicit USM Utility Functions
+'''
+==== malloc
+[source,cpp]
+----
+void* sycl_malloc_device(size_t size);
+----
+
+Parameters:: `size_t size` - number of bytes to allocate
+Return value:: Returns a pointer to the newly allocated memory on the `device` selected by the default selector on success.   Memory allocated by `sycl_malloc_device` must be deallocated with `sycl_free` to avoid memory leaks. On failure, returns `nullptr`.
+
+'''
+
+[source,cpp]
+----
+void* sycl_malloc_device(size_t size, const sycl::device& dev);
```

Comment:
All your functions start with `sycl_`. What about using some C++ namespaces?
Otherwise the API looks very C like. Which is nice to have the code usable from C applications.
But do you plan to add some higher-level modern C++ fancy API?

Reply:
Yes, that part have a very C-like API by design in order to minimize spec changes.  Building a C++-style allocator on top of those functions should be very trivial to do in an implementation.